sequencing-server/src/utils/hasura.ts

```diff
@@ -0,0 +1,108 @@
+ * @returns The AuthSessionVeriables of the user sending the request.
+ */
+export function getHasuraSession(
+  authSessionVariables: AuthSessionVariables | undefined,
+  authHeader: string | undefined,
+): AuthSessionVariables {
+  if (authSessionVariables !== undefined) {
+    return authSessionVariables;
+  } else if (authHeader !== undefined) {
+    return (jwtDecode(authHeader) as Jwt)['https://hasura.io/jwt/claims'];
+  }
+
+  throw new Error('Could not determine the user sending the request');
+}
+
+/**
+ * A helper function to get the username of the user who sent the request.
+ *
+ * @param authSessionVariables If the request has gone through a hasura action this object will be populated.
+ * @param authHeader If the user sends a POST request directly to the endpoint with a bearer token this will be populated.
+ * @returns The username of the user sending the request.
+ */
+export function getUsername(
+  authSessionVariables: AuthSessionVariables | undefined,
+  authHeader: string | undefined,
+): string {
+  if (authSessionVariables !== undefined) {
+    return authSessionVariables['x-hasura-user-id'];
+  } else if (authHeader !== undefined) {
+    return (jwtDecode(authHeader) as Jwt).username;
+  }
+
+  throw new Error('Could not determine the user sending the request');
+}
+
+/**
+ * This function checks to see if request coming from the Hasura action is being executed
+ * by a user with enough permissions. To check this we need to call a GQL query with the
+ * action key and the user session. If we ever merge the databases this should be changed
+ * to call the db function that does the permission checking directly.
+ *
+ * Some endpoints don't need to be checked which is why we have the mappings of endpoints
+ * to actions to check.
+ *
+ * @param url The endpoint that we're checking the permissions for.
+ * @param graphqlClient The GQL Client we're using to make the permission check.
+ * @returns True if the user is able to call the given action, false otherwise.
+ */
+export async function canUserPerformAction(
+  url: string,
+  graphqlClient: GraphQLClient,
+  hasuraSession: AuthSessionVariables,
+): Promise<boolean> {
+  if (url in ENDPOINTS_TO_ACTION_KEY) {
+    const role = hasuraSession['x-hasura-role'];
+
+    // The aerie_admin role always has NO_CHECK permissions on all actions.
+    if (role === 'aerie_admin') {
+      return true;
+    }
+
+    const permissionCheckQuery = await graphqlClient.request(
+      gql`
+        query checkUserPermissions($role: user_roles_enum!, $key: String!) {
+          user_role_permission_by_pk(role: $role) {
+            action_permissions(path: $key)
+          }
+        }
+      `,
+      {
+        role,
+        key: ENDPOINTS_TO_ACTION_KEY[url],
+      },
+    );
+
+    if (permissionCheckQuery.user_role_permission_by_pk.action_permissions === null) {
+      return false;
+    }
+  }
+
+  return true;
```

Comment:
Just a reminder that this doesn't yet sufficiently check if the user has permission to run the function (although it seems like you're still working on this part)

Reply:
I didn't see any keys (for the seq endpoints) that indicated how I should do the checking so I assumed that a `null` corresponded to not enough permissions, is there something else I should be checking here?

Reply:
While `null` does correspond to "no permission", the general keys are usable on **all** endpoints, **unless there is a reason that a specific key does not make sense on a specific endpoint**.
If that is the case for any key, then specify *why* it cannot be applied so that it can be documented. Additionally, the endpoint needs to _reject_ the invalid permissions and _check_ that user satisfies the valid permissions.

Under the current implementation, the following unauthorized transactions would not be rejected:
1. The role `user` has the permission `{expand_all_activities: PLAN_OWNER_SOURCE}` and a User with that role attempts to run `expand_all_activities`. Expected behavior: the request is rejected because the permission rule to check is invalid (PLAN_OWNER_SOURCE is a merge permission and `expand_all_activities` is not a merge action)
2. The role `user` has the permission `{expand_all_activities: PLAN_OWNER}` and a User with that role _who does not owner the plan that was simulated_ attempts to run `expand_all_activities`. Expected behavior: the requested is rejected because the User does not satisfy the permission rule (they do not own the Plan).